clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -659,4 +662,24 @@ public static ClassLoader getContextOrKafkaClassLoader() {
             return cl;
     }
 
+    /**
+     * Attempts to move source to target atomically and falls back to a non-atomic move if it fails.
+     *
+     * @throws IOException if both atomic and non-atomic moves fail
+     */
+    public static void atomicMoveWithFallback(Path source, Path target) throws IOException {
+        try {
+            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException outer) {
+            try {
+                Files.move(source, target, StandardCopyOption.REPLACE_EXISTING);
+                log.debug("Non-atomic move of " + source + " to " + target + " succeeded after atomic move failed due to "
+                        + outer.getMessage());
+            } catch (IOException inner) {
+                inner.addSuppressed(outer);
+                throw inner;
```

Comment:
Does it actually get logged? I don't know that it adds much value (anyone encountering an exception hitting these lines can easily see that there would have been a previous one), but it also doesn't hurt to include it.